### PR TITLE
MM-58525 Fix upload file permissions

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
Both `create_post` and `upload_file` are default permissions on a ChannelUser and ChannelGuest role. The `create_post` permission can be removed from either a ChannelUser or a ChannelGuest in the System Console. However, there is not a way to remove the `upload_file` permission.

The `upload_file` is checked when a user attempts to upload a file via API. Therefore, a user with `create_post` removed can still upload a file via API. 

This PR ties the two permission together for the ChannelUser and ChannelGuest roles. It also createa a database migration to remove `upload_file` where the `create_post` permission does not exist.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-58525

#### Screenshots
Screenshot 2024-06-11 at 2 51 09 PM
Screenshot 2024-06-11 at 2 50 41 PM

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Review by sbishel — COMMENTED** [review]
Batching the data migrations are probably overkill. There would only be 2 rows per Team Schema Override, so it would take 50 schema overrides (very doubtful) and all wold have to have "create_post" removed for the loop to occur.

But...the [documentation](https://developers.mattermost.com/contribute/more-info/server/schema-migration-guide/#how-do-i-run-update-statements-in-data-migrationsdenormalizations) provided the code sample, so it was easy enough to add.

File: server/channels/db/migrations/mysql/000122_remove_upload_file_permission.up.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
+CREATE PROCEDURE RemoveUuploadFilePermission()
+BEGIN
+    updateRoles: LOOP
+        -- update affected rows
+        UPDATE Roles 
+            SET Permissions = REGEXP_REPLACE(Permissions, 'upload_file[[:space:]|?]', '')
+            WHERE Permissions like '%upload_file%' and Permissions not REGEXP 'create_post[[:space:]|?]'
```
Comment (sbishel) [inline server/channels/db/migrations/mysql/000122_remove_upload_file_permission.up.sql:7]: There is a permission `create_post_public`. Need to use regex, to ensure this one is not selected.

File: server/channels/db/migrations/postgres/000122_remove_upload_file_permission.up.sql [hunk 2]
```diff
@@ -0,0 +1,31 @@
+DO $$
+<<remove_upload_file_permission>>
+DECLARE
```
Comment (agnivade) [inline server/channels/db/migrations/postgres/000122_remove_upload_file_permission.up.sql:3]: Glad to see my stored proc being used :). Just wanted to confirm, you have tested this out and it works as expected right?

**Review by agnivade — APPROVED** [review]

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1: We'll probably want to update the documentation as a consequence of this PR in places [like this one](https://docs.mattermost.com/onboard/advanced-permissions.html#read-only-channels)

**Review by lindy65 — APPROVED** [review]
Thanks @sbishel - looks good to me 👍

**Comment by amyblais** [comment]
@sbishel Can you help submit a manual cherry-pick PR for ``release-9.10`` branch?

**Comment by amyblais** [comment]
@sbishel Can you help submit manual cherry-picks PR for ``release-9.10``, ``release-9.9``, ``release-9.8``, and ``release-9.5`` branches?

**Pull request merged**